### ISSUE TITLE
destroy method in yii.gridView.js doesn't work as expected

### ISSUE THREAD

**Issue by arogachev** [yiisoft/yii2#13231]
Related issue for `yii.captcha.js`: #13159.

Let's take a look at the current implementation of `destroy` method in `yii.gridView.js`.

```
destroy: function () {
    return this.each(function () {
        $(window).unbind('.yiiGridView');
        $(this).removeData('yiiGridView');
    });
},
```

Removing of event handlers is incorrect:

- `unbind()` will have no effect because of the `$(window)`.
- Usage of [unbind](http://api.jquery.com/unbind/) is discouraged since jQuery 1.7 and it's deprecated starting from version 3.0.
- We can use `$(document)` instead of a `$(window)` but this contradicts with multiple grid views initialization concept. Destroying single grid view should not affect others.
- Because of destroying single grid view, loop and wrapping `this` with `$()` is not necessary.
- `beforeFilter` and `afterFilter` event handlers are not included (they don't have `yiiGridView` namespace), with each repeated initialization we can get +1 extra call:
- With destroying of a single grid view event handlers for filter, checking 1 row / all rows are not detached because they were attached to different elements, not to GridView itself. For example. event handler for checking all rows:

```js
 $(document).off('click.yiiGridView', checkAll).on('click.yiiGridView', checkAll, function () {
    $grid.find(inputs + ":enabled").prop('checked', this.checked);
});
```

`checkAll` is declared above:

```js
var checkAll = "#" + id + " input[name='" + options.checkAll + "']";
```

and can't be for example `#w0 input[name="selection_all"]`.

And finally removing of grid view setings is incorrect:

```js
$(this).removeData('yiiGridView');
```

I guess it was copy pasted from `yii.captcha.js` where settings stored in the `data` attribute of according HTML element. In `yii.gridView.js` settings are stored in `gridData` container (private variable), so correct way of deleting is just:

```js
delete gridData[id];
```

**Comment by samdark** [comment]
Is it fixed now?

**Comment by arogachev** [comment]
@samdark Yes, it's fixed in #13230.
